sympy/codegen/ast.py

```diff
@@ -335,6 +335,11 @@ def kwargs(self, exclude=(), apply=None):
         else:
             return kwargs
 
+    def func(self, *args, **kwargs):
+        if self.is_Atom and not kwargs:
+            return self
+        else:
+            return super().func(*args, **kwargs)
```

Comment:
The kwargs thing here is an antipattern as far as I'm concerned. I'm not sure how exactly this class is used but can we not make it so that it just stores everything in args?

Reply:
Yes, the `str` argument (non-basic) could be stored as a `String`. These classes (`Token` and its subclasses) were designed with the idea that you could both have `args` arguments and non `args` arguments. Note that the string is stored in a non `_` argument, so changing the type from `str` to `String` should be considered an API change (hence why I tried to circumvent this)

Reply:
Maybe we can store String in `_args` even if there is also a str stored on the instance.

Reply:
It is certainly possible, but `Token` has been explicitly designed with allowing arguments that are not stored in `_args`, : 
```python
class String(Token):
    """ SymPy object representing a string.
    Atomic object which is not an expression (as opposed to Symbol).
    Parameters
    ==========
    text : str
    Examples
    ========
    >>> from sympy.codegen.ast import String
    >>> f = String('foo')
    >>> f
    foo
    >>> str(f)
    'foo'
    >>> f.text
    'foo'
    >>> print(repr(f))
    String('foo')
    """
    __slots__ = ('text',)
    not_in_args = ['text']
    is_Atom = True
```
`Token`'s __new__ takes care of everything, the `String` class doesn't have any of that logic directly. I would assume that `Str` ( `Atom` that can contain `str`) didn't exist when this was made.

Also, assume we use `Str` to store the `text` in args, then what happens when `func` is used, now it is being given a `Str` instead of a `str`, sure we can handle this for `str` but this is expected to work for everything else too (considering also downstream projects making use of the currently provided interface).

Note that the kwargs behaviour is currently actually being tested for (see the line above the line introduced in the test script in this commit)

Reply:
```python
def test_String():
    st = String('foobar')
    assert st.is_Atom
    assert st == String('foobar')
    assert st.text == 'foobar'
    assert st.func(**st.kwargs()) == st <--was already in the test
    assert st.func(*st.args) == st <-- newly added
```

Reply:
May @bjodah can help here.

I'm wondering if there is a backwards compatible way to change this so that we can satisfy the func-args invariant:
```
obj.func(*obj,args) == obj  # for all any obj that is Basic
```
We can't have `not_in_args` if that means that the object is defined by things that are not in the `args`.

Reply:
The current suggestion in the commit would:
- work with `st.funct(**st.kwargs()) == st`
- work if any `Token` instance has only `args` arguments
- work if any `Token` instance has only non `args` arguments (new, currently only passes the arg invariance check since it has no arguments)
- still does not work if there is a mix of `args` and non `args` arguments

Reply:
> ange this so that we can satisfy the func-args invariant:

Well, func could "zip" together `args` (newly supplied) and `not_in_args` (the values stored in the instance), in which case `not_in_args` would act "atomic" in the sense that they can't be modified. An if statement for keyword arguments can then be used to revert to the kwarg invariant

Reply:
We can do lots of manipulations in `func` but I would prefer not to do any unless there is a good reason.

Reply:
I vaguely remember that we needed code generation to work even with non-Basic objects (I think `Matrix`, at least at that time was not a subclass of `Basic`), and using non-Sympy types were problematic. Hence the complicated design of `Token`, though most of its complexity deals with pretty-printing its repr and allowing it to be constructed using kwargs, example from the docs of `Element` (which uses Tuple's __new__ impl and only declares types of its arguments in its class body):
```python
>>> ccode(Element('x', 'ijk', strides='lmn', offset='o'))
'x[i*l + j*m + k*n + o]'
```
So personally I like kwargs in `__new__` (since multiple positional arguments are in general bad unless the order is apparent), as for `func` I think it's fine to stop supporting it (if we can get all values currently stored in __slots__ moved over).

String only takes a single argument, so kwargs here are superfluous.

If Token could be made to support the `.func(*args)` invariance in general that's great. And if all values in `__slots__` could go into args, that's also great (though I wonder if whatever reason we had not to in the first place isn't still an issue...).

Reply:
For subclasses from `Basic`, `func(*args)` is demanded to be supported (e.g. `subs` uses it). In the current design, only `Atom`s can have non-`Basic` arguments that have to be stored (it is fine if it is used to decide what object to create or if it is converted to a `Basic` somehow).

Reply:
This is always the hard part. It's difficult to make assumptions about downstream code but we have to,

Reply:
My assumption for downstream code would be that people have `Token` subclasses that store non-`Basic` objects that they want printed code for, since this is currently explicitly supported. Having some "manipulations" in `func` seems like a reasonable trade-off if it leads to supporting arg invariance and be backwards compatible, both of which are formal requirements.

Reply:
It looks like `String` is the only class that uses `not_in_args` in the SymPy codebase. Maybe if it was an Atom subclass then it could inherit `func` from `Atom`.

Reply:
It could, but it would not be backwards compatible with  `obj.func(**obj.kwargs())` (hence the `and not kwargs` in the proposed `func`)

Reply:
It would be compatible if the `kwargs` method was changed to return an empty dict.

Reply:
I have given this a try, perhaps in a later PR the arg invariance can be solved more generically